File: app/widgets/appLauncher/AppLauncher.java
package appLauncher;

import de.jensd.fx.glyphs.fontawesome.FontAwesomeIcon;
import java.io.File;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicLong;
import kotlin.sequences.Sequence;
import sp.it.pl.gui.objects.grid.GridFileThumbCell;
import sp.it.pl.gui.objects.grid.GridFileThumbCell.Loader;
import sp.it.pl.gui.objects.grid.GridView;
import sp.it.pl.gui.objects.grid.GridView.CellSize;
import sp.it.pl.gui.objects.hierarchy.Item;
import sp.it.pl.gui.objects.placeholder.Placeholder;
import sp.it.pl.layout.widget.ExperimentalController;
import sp.it.pl.layout.widget.Widget;
import sp.it.pl.layout.widget.controller.LegacyController;
import sp.it.pl.layout.widget.controller.SimpleController;
import sp.it.pl.main.Widgets;
import sp.it.util.Sort;
import sp.it.util.access.V;
import sp.it.util.access.VarEnum;
import sp.it.util.access.fieldvalue.FileField;
import sp.it.util.async.executor.FxTimer;
import sp.it.util.conf.Config.VarList;
import sp.it.util.conf.Config.VarList.Elements;
import sp.it.util.conf.IsConfig;
import sp.it.util.dev.Dependency;
import sp.it.util.file.FileSort;
import sp.it.util.file.FileType;
import sp.it.util.ui.Resolution;
import static de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIcon.FOLDER_PLUS;
import static java.util.stream.Collectors.toList;
import static javafx.scene.input.KeyCode.ENTER;
import static javafx.scene.input.MouseButton.SECONDARY;
import static javafx.util.Duration.millis;
import static kotlin.streams.jdk8.StreamsKt.asSequence;
import static sp.it.pl.gui.objects.grid.GridView.CellSize.NORMAL;
import static sp.it.pl.gui.objects.grid.GridView.SelectionOn.KEY_PRESS;
import static sp.it.pl.gui.objects.grid.GridView.SelectionOn.MOUSE_CLICK;
import static sp.it.pl.gui.objects.grid.GridView.SelectionOn.MOUSE_HOVER;
import static sp.it.pl.layout.widget.Widget.Group.OTHER;
import static sp.it.pl.main.AppDragKt.installDrag;
import static sp.it.pl.main.AppExtensionsKt.scaleEM;
import static sp.it.pl.main.AppKt.APP;
import static sp.it.util.Sort.ASCENDING;
import static sp.it.util.async.AsyncKt.FX;
import static sp.it.util.async.AsyncKt.oneTPExecutor;
import static sp.it.util.async.AsyncKt.runFX;
import static sp.it.util.async.AsyncKt.runIO;
import static sp.it.util.async.executor.FxTimer.fxTimer;
import static sp.it.util.file.FileSort.DIR_FIRST;
import static sp.it.util.file.FileType.FILE;
import static sp.it.util.functional.Util.by;
import static sp.it.util.functional.Util.list;
import static sp.it.util.functional.UtilKt.consumer;
import static sp.it.util.functional.UtilKt.runnable;
import static sp.it.util.math.UtilKt.max;
import static sp.it.util.reactive.UtilKt.attach1IfNonNull;
import static sp.it.util.reactive.UtilKt.sync1IfInScene;
import static sp.it.util.system.EnvironmentKt.chooseFile;
import static sp.it.util.system.EnvironmentKt.open;

@Widget.Info(
    author = "Martin Polakovic",
    name = Widgets.APP_LAUNCHER,
    description = "Launches programs",
    version = "0.8.0",
    year = "2016",
    group = OTHER
)
@ExperimentalController(reason = "DirView widget could be improved to be fulfill this widget's purpose. Also needs better UX.")
@LegacyController
public class AppLauncher extends SimpleController {

    private static final double CELL_TEXT_HEIGHT = 40;

    @IsConfig(name = "Location", info = "Add program")
    final VarList<File> files = new VarList<>(File.class, Elements.NOT_NULL);
    List<File> filesMaterialized = new ArrayList<>(files.list);
    @IsConfig(name = "Thumbnail size", info = "Size of the thumbnail.")
    final V<CellSize> cellSize = new V<>(NORMAL).initAttachC(v -> applyCellSize());
    @IsConfig(name = "Thumbnail size ratio", info = "Size ratio of the thumbnail.")
    final V<Resolution> cellSizeRatio = new V<>(Resolution.R_4x5).initAttachC(v -> applyCellSize());

    private final GridView<Item, File> grid = new GridView<>(File.class, v -> v.value, cellSize.get().width,cellSize.get().width*cellSizeRatio.get().ratio +CELL_TEXT_HEIGHT,5,5);
    private final Loader imageLoader = new Loader(oneTPExecutor());
    boolean initialized = false;
    private volatile boolean isResizing = false;
	private final AtomicLong visitId = new AtomicLong(0);
    private final Placeholder placeholder = new Placeholder(
        FOLDER_PLUS, "Click to add launcher or drag & drop a file",
        runnable(() ->
            chooseFile("Choose program or file", FILE, APP.DIR_HOME, root.getScene().getWindow())
				.ifOkUse(files.list::setAll)
        )
    );

    @IsConfig(name = "Sort", info = "Sorting effect.")
    final V<Sort> sort = new V<>(ASCENDING).initAttachC(v -> applySort());
    @IsConfig(name = "Sort file", info = "Group directories and files - files first, last or no separation.")
    final V<FileSort> sort_file = new V<>(DIR_FIRST).initAttachC(v -> applySort());
    @IsConfig(name = "Sort by", info = "Sorting criteria.")
    final V<FileField<?>> sortBy = new VarEnum<>(FileField.NAME, () -> FileField.FIELDS).initAttachC(v -> applySort());
    @Dependency("name: used by reflection")
    @IsConfig(name = "Close on launch", info = "Close this widget when it launches a program.")
    final V<Boolean> closeOnLaunch = new V<>(false);
    @Dependency("name: used by reflection")
    @IsConfig(name = "Close on right click", info = "Close this widget when right click is detected.")
    final V<Boolean> closeOnRightClick = new V<>(false);

    public AppLauncher(Widget widget) {
        super(widget);
        root.setPrefSize(scaleEM(1000), scaleEM(700));

        files.onListInvalid(list -> filesMaterialized = new ArrayList<>(files.list));
        files.onListInvalid(list -> visit());
        files.onListInvalid(list -> placeholder.show(root, list.isEmpty()));
        grid.search.field = FileField.PATH;
        grid.primaryFilterField = FileField.NAME_FULL;
        grid.selectOn.addAll(list(KEY_PRESS, MOUSE_CLICK, MOUSE_HOVER));
        grid.setCellFactory(grid -> new Cell());
        root.getChildren().add(grid);
        placeholder.showFor(root);

        // delay cell loading when content is being resized (increases resize performance)
        double delay = 200; // ms
        FxTimer resizeTimer = fxTimer(millis(delay), 1, runnable(() -> isResizing = false));
        grid.widthProperty().addListener((o,ov,nv) -> isResizing = true);
        grid.heightProperty().addListener((o,ov,nv) -> isResizing = true);
        grid.widthProperty().addListener((o,ov,nv) -> resizeTimer.start(millis(300)));
        grid.heightProperty().addListener((o,ov,nv) -> resizeTimer.start(millis(300)));
        grid.setOnKeyPressed(e -> {
            if (e.getCode()==ENTER) {
                Item si = grid.selectedItem.get();
                if (si!=null) doubleClickItem(si);
            }
        });
        grid.setOnMouseClicked(e -> {
            if (e.getButton()==SECONDARY && closeOnRightClick.get())
                widget.uiTemp.dispose();  // TODO: this is for overlayPane interop, move out of here and use hide() instead of dispose()
        });

        installDrag(
            root, FontAwesomeIcon.PLUS_SQUARE_ALT, "Add launcher",
            e -> e.getDragboard().hasFiles(),
            consumer(e -> files.list.addAll(e.getDragboard().getFiles()))
        );

        sync1IfInScene(root, runnable(() -> {
            initialized = true;
            applyCellSize();
            visit();
        }));

        onClose.plusAssign(runnable(() -> imageLoader.shutdown()));
    }

    @Override
    public void focus() {
        attach1IfNonNull(grid.skinProperty(), consumer(skin -> grid.implGetSkin().requestFocus()));
    }

    private void visit() {
        if (!initialized) return;
        Item item = new TopItem();
	    visitId.incrementAndGet();
        runIO(() ->
            item.children().stream().sorted(buildSortComparator()).collect(toList())
        ).useBy(FX, cells -> {
            grid.getItemsRaw().setAll(cells);

            grid.implGetSkin().setPosition(max(item.lastScrollPosition, 0.0));
            grid.requestFocus();    // fixes focus problem
        });
    }

    private void doubleClickItem(Item i) {
        if (closeOnLaunch.get()) {
            widget.uiTemp.dispose();
            runFX(millis(250), () -> open(i.value));
        } else {
            open(i.value);
        }
    }

    /** Resorts grid's items according to current sort criteria. */
    private void applySort() {
        grid.getItemsRaw().sort(buildSortComparator());
    }

    void applyCellSize() {
        grid.setCellWidth(cellSize.get().width);
        grid.setCellHeight((cellSize.get().width/cellSizeRatio.get().ratio)+CELL_TEXT_HEIGHT);
    }

    private Comparator<Item> buildSortComparator() {
        Sort sortHetero = sort_file.get().getSort(), // sorts Files to files and directories
             sortHomo = sort.get(); // sorts each group separately
        FileField<?> field = sortBy.get(); // pre-compute once for consistency and performance
        Comparator<Item> cmpHetero = sortHetero.of(by(i -> i.valType)),
                         cmpHomo = sortHomo.of(by(i -> i.value, field.comparator()));
        return cmpHetero.thenComparing(cmpHomo);
    }

    /**
     * Graphics representing the file. Cells are virtualized just like ListView or TableView does
     * it, but both vertically & horizontally. This avoids loading all files at once and allows
     * unlimited scaling.
     */
    private class Cell extends GridFileThumbCell {
        public Cell() {
            super(imageLoader);
        }

        @Override
        protected double computeCellTextHeight() {
            return CELL_TEXT_HEIGHT;
        }

        @Override
        protected void onAction(Item i, boolean edit) {
            doubleClickItem(i);
        }
    }

    private static class FItem extends Item {

        public FItem(Item parent, File value, sp.it.util.file.FileType type) {
            super(parent, value, type);
        }

        @Override
        protected FItem createItem(Item parent, File value, sp.it.util.file.FileType type) {
			return getPortableAppExe(value, type)
				.map(f -> new FItem(parent, getPortableAppExe(value, type).orElse(null), FileType.FILE))
				.orElseGet(() -> new FItem(parent, value, type));
        }

    }
    private class TopItem extends FItem {

        public TopItem() {
            super(null,null,null);
        }

	    @Override
	    protected Sequence<File> childrenFiles() {
		    return asSequence(filesMaterialized.stream().distinct());
	    }

	    @Override
        protected File getCoverFile() {
            return null;
        }
    }

    public static Optional<File> getPortableAppExe(File f, FileType type) {
    	return type==FileType.DIRECTORY
					? Optional.of(new File(f, f.getName() + ".exe"))
					: Optional.empty();
    }
}

File: app/widgets/appLauncher/AppLauncher.kt
package appLauncher

import com.sun.jna.Native
import com.sun.jna.platform.win32.Shell32
import com.sun.jna.platform.win32.ShlObj
import com.sun.jna.platform.win32.WinDef
import javafx.scene.input.KeyCode.ENTER
import javafx.scene.input.KeyEvent.KEY_PRESSED
import javafx.scene.input.MouseButton.SECONDARY
import javafx.scene.input.MouseEvent.MOUSE_CLICKED
import javafx.util.Callback
import sp.it.pl.gui.objects.grid.GridFileThumbCell
import sp.it.pl.gui.objects.grid.GridFileThumbCell.Loader
import sp.it.pl.gui.objects.grid.GridView
import sp.it.pl.gui.objects.grid.GridView.CellSize.NORMAL
import sp.it.pl.gui.objects.grid.GridView.SelectionOn.KEY_PRESS
import sp.it.pl.gui.objects.grid.GridView.SelectionOn.MOUSE_CLICK
import sp.it.pl.gui.objects.grid.GridView.SelectionOn.MOUSE_HOVER
import sp.it.pl.gui.objects.hierarchy.Item
import sp.it.pl.layout.widget.ExperimentalController
import sp.it.pl.layout.widget.Widget
import sp.it.pl.layout.widget.Widget.Group.OTHER
import sp.it.pl.layout.widget.controller.SimpleController
import sp.it.pl.main.IconFA
import sp.it.pl.main.Widgets
import sp.it.pl.main.installDrag
import sp.it.pl.main.scaleEM
import sp.it.pl.main.withAppProgress
import sp.it.util.Sort.ASCENDING
import sp.it.util.access.fieldvalue.CachingFile
import sp.it.util.access.fieldvalue.FileField
import sp.it.util.async.IO
import sp.it.util.async.future.Fut.Companion.fut
import sp.it.util.async.oneTPExecutor
import sp.it.util.async.runFX
import sp.it.util.async.runIO
import sp.it.util.collections.materialize
import sp.it.util.collections.setTo
import sp.it.util.conf.IsConfig
import sp.it.util.conf.c
import sp.it.util.conf.cList
import sp.it.util.conf.cv
import sp.it.util.conf.only
import sp.it.util.conf.values
import sp.it.util.dev.Dependency
import sp.it.util.file.FileSort.DIR_FIRST
import sp.it.util.file.FileType
import sp.it.util.file.FileType.FILE
import sp.it.util.file.div
import sp.it.util.functional.net
import sp.it.util.functional.nullsLast
import sp.it.util.functional.toUnit
import sp.it.util.inSort
import sp.it.util.math.max
import sp.it.util.reactive.attach1IfNonNull
import sp.it.util.reactive.onChange
import sp.it.util.reactive.onEventDown
import sp.it.util.reactive.sync1IfInScene
import sp.it.util.system.Os
import sp.it.util.system.open
import sp.it.util.ui.Resolution
import sp.it.util.ui.image.FitFrom
import sp.it.util.ui.lay
import sp.it.util.ui.prefSize
import sp.it.util.ui.x
import sp.it.util.units.millis
import sp.it.util.validation.Constraint.FileActor
import java.io.File
import java.util.concurrent.atomic.AtomicLong

@Widget.Info(
        author = "Martin Polakovic",
        name = Widgets.APP_LAUNCHER,
        description = "Application menu and launcher",
        version = "0.8.0",
        year = "2016",
        group = OTHER
)
@ExperimentalController(reason = "DirView widget could be improved to be fulfill this widget's purpose. Also needs better UX.")
class AppLauncher(widget: Widget): SimpleController(widget) {

    @IsConfig(name = "Location", info = "Add program")
    private val files by cList<File>().only(FileActor.DIRECTORY)
    private var filesMaterialized = files.materialize()

    @IsConfig(name = "Thumbnail size", info = "Size of the thumbnail.")
    private val cellSize by cv(NORMAL) attach { applyCellSize() }
    @IsConfig(name = "Thumbnail size ratio", info = "Size ratio of the thumbnail.")
    private val cellSizeRatio by cv(Resolution.R_1x1) attach { applyCellSize() }
    @IsConfig(name = "Thumbnail fit image from", info = "Determines whether image will be fit from inside or outside.")
    private val fitFrom by cv(FitFrom.OUTSIDE)

    private val grid = GridView<Item, File>(File::class.java, { it.value }, cellSize.value.width, cellSize.value.width/cellSizeRatio.value.ratio+CELL_TEXT_HEIGHT, 5.0, 5.0)
    private val imageLoader = Loader(oneTPExecutor())
    private val visitId = AtomicLong(0)

    private var item: Item? = null   // item, children of which are displayed

    @IsConfig(name = "Sort", info = "Sorting effect.")
    private val sort by cv(ASCENDING) attach { applySort() }
    @IsConfig(name = "Sort first", info = "Group directories and files - files first, last or no separation.")
    private val sortFile by cv(DIR_FIRST) attach { applySort() }
    @IsConfig(name = "Sort seconds", info = "Sorting criteria.")
    private val sortBy by cv<FileField<*>>(FileField.NAME).values(FileField.FIELDS) attach { applySort() }
    @Dependency("Must be Config. Accessed in the application by name")
    @IsConfig(name = "Close on launch", info = "Close this widget when it launches a program.")
    private var closeOnLaunch by c(false)
    @IsConfig(name = "Close on right click", info = "Close this widget when right click is detected.")
    private var closeOnRightClick by c(false)

    init {
        root.prefSize = 1000.scaleEM() x 700.scaleEM()

        grid.search.field = FileField.PATH
        grid.primaryFilterField = FileField.NAME_FULL
        grid.selectOn setTo listOf(KEY_PRESS, MOUSE_CLICK, MOUSE_HOVER)
        grid.cellFactory = Callback { Cell() }
        root.lay += grid

        grid.onEventDown(KEY_PRESSED, ENTER, false) { e ->
            grid.selectedItem.value?.let {
                doubleClickItem(it)
                e.consume()
            }
        }
        grid.onEventDown(MOUSE_CLICKED, SECONDARY, false) {
            if (closeOnRightClick) {
                doubleClickItem(null)
                it.consume()
            }
        }

        installDrag(
                root, IconFA.PLUS_SQUARE_ALT, "Add launcher",
                { e -> e.dragboard.hasFiles() },
                { e -> files += e.dragboard.files }
        )

        files.onChange { filesMaterialized = files.materialize() }
        files.onChange { visit() }
        onClose += { disposeItem() }
        onClose += { imageLoader.shutdown() }

        root.sync1IfInScene {
            applyCellSize()
        }
    }

    override fun focus() = grid.skinProperty().attach1IfNonNull { grid.implGetSkin().requestFocus() }.toUnit()

    private fun visit() {
        disposeItem()
        val i = TopItem()
        item = i
        visitId.incrementAndGet()
        runIO {
            i.children().sortedWith(buildSortComparator())
        }.withAppProgress(
                widget.custom_name.value+": Fetching view"
        ) ui {
            grid.itemsRaw setTo it
            grid.implGetSkin().position = i.lastScrollPosition max 0.0
            grid.requestFocus()
        }
    }

    private fun disposeItem() {
        item?.dispose()
        item = null
    }

    private fun doubleClickItem(i: Item?) {
        if (closeOnLaunch) {
            widget.uiTemp.dispose() // TODO: this is for overlayPane inter-op, move out of here and use hide() instead of dispose()
            runFX(250.millis) { i?.value?.open() }
        } else {
            i?.value?.open()
        }
    }

    private fun applyCellSize(width: Double = cellSize.value.width, height: Double = cellSize.value.width/cellSizeRatio.value.ratio) {
        grid.setCellSize(width, height+CELL_TEXT_HEIGHT)
        visit()
    }

    private fun applySort() {
        fut(grid.itemsRaw.materialize()).then(IO) {
            it.sortedWith(buildSortComparator())
        } ui {
            grid.itemsRaw setTo it
        }
    }

    private fun buildSortComparator() = compareBy<Item> { 0 }
            .thenBy { it.valType }.inSort(sortFile.value.sort)
            .thenBy(sortBy.value.comparator<File> { it.inSort(sort.value).nullsLast() }) { it.value }

    private inner class Cell: GridFileThumbCell(imageLoader) {

        override fun computeCellTextHeight() = CELL_TEXT_HEIGHT

        override fun onAction(i: Item, edit: Boolean) = doubleClickItem(i)

    }

    private open class FItem(parent: Item?, value: File?, type: FileType?): Item(parent, value, type) {

        override fun createItem(parent: Item, value: File, type: FileType) = null
                ?: value.getPortableAppExe(type)?.net { FItem(parent, it, FILE) }
                ?: FItem(parent, value, type)

    }

    private inner class TopItem: FItem(null, null, null) {

        override fun childrenFiles() = (filesMaterialized.asSequence()+startMenuPrograms()).distinct()

        override fun getCoverFile() = null

    }

    companion object {

        private const val CELL_TEXT_HEIGHT = 20.0

        fun File.getPortableAppExe(type: FileType) = if (type==FileType.DIRECTORY) File(this, "$name.exe") else null

        fun startMenuPrograms(): Sequence<File> = when (Os.current) {
            Os.WINDOWS -> windowsAppData().walk()
                    .filter { '.' in it.name && !it.name.contains("uninstall", true) }
                    .map { CachingFile(it) }
            else -> sequenceOf()
        }

        fun windowsAppData(): File {
            val pszPath = CharArray(WinDef.MAX_PATH)
            Shell32.INSTANCE.SHGetFolderPath(null, ShlObj.CSIDL_APPDATA, null, ShlObj.SHGFP_TYPE_CURRENT, pszPath)
            val path = Native.toString(pszPath)
            return File(path)/"Microsoft"/"Windows"/"Start Menu"/"Programs"
        }

    }
}